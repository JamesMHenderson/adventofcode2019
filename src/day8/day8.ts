type Size = {
    x: number;
    y: number;
};

const createImageLayerString = (input: string, { x, y }: Size): Array<string> => {
    const layers = [];
    const layerSize = x * y;

    let i = 0;

    while (i < input.length) {
        layers.push(input.substr(i, layerSize));
        i += layerSize;
    }

    return layers;
};

const checkLayers = (input: string, size: Size): number => {
    const layers = createImageLayerString(input, size);

    const fewest0 = layers.reduce((acc, value) => {
        if (value.split('0').length < acc.split('0').length) {
            return value;
        } else {
            return acc;
        }
    });

    return Array.from(fewest0).filter(x => x === '1').length * Array.from(fewest0).filter(x => x === '2').length;
};

const createImageLayers = (input: string, { x, y }: Size): Array<Array<string>> => {
    const layers: Array<Array<string>> = [];
    const layerSize = x * y;

    let i = 0;

    while (i < input.length) {
        let j = 0;
        const layerInput = input.substr(i, layerSize);
        const layer = [];
        while (j < layerInput.length) {
            layer.push(layerInput.substr(j, x));
            j += x;
        }
        layers.push(layer);
        i += layerSize;
    }

    return layers;
};

const createImage = (input: string, size: Size): Array<string> => {
    const layers = createImageLayers(input, size);
    const image = [];

    for (let j = 0; j < size.y; j++) {
        let layerRow = '';
        for (let i = 0; i < size.x; i++) {
            for (let k in layers) {
                if (layers[k][j][i] !== '2') {
                    layerRow += layers[k][j][i] === '0' ? '.' : '#';
                    break;
                }
            }
        }
        image.push(layerRow);
    }

    return image;
};

export { createImageLayers, checkLayers, createImage };
