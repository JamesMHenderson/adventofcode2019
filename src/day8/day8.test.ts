import { createImageLayers, checkLayers, createImage } from './day8';
import input from './input.txt';

describe('day 8', () => {
    describe('Part 1', () => {
        test('example', () => {
            expect(checkLayers('123456789012', { x: 3, y: 2 })).toBe(1);
        });

        test('challenge', () => {
            expect(checkLayers(input.trim(), { x: 25, y: 6 })).toBe(1690);
        });
    });

    describe('Part 2', () => {
        test('example', () => {
            expect(createImage('0222112222120000', { x: 2, y: 2 })).toMatchSnapshot();
        });

        test('challenge', () => {
            expect(createImage(input.trim(), { x: 25, y: 6 })).toMatchSnapshot();
        });
    });
});
