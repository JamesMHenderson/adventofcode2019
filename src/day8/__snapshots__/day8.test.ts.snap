// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`day 8 Part 2 challenge 1`] = `
Array [
  "####.###..####.#..#.###..",
  "...#.#..#....#.#..#.#..#.",
  "..#..#..#...#..#..#.###..",
  ".#...###...#...#..#.#..#.",
  "#....#....#....#..#.#..#.",
  "####.#....####..##..###..",
]
`;

exports[`day 8 Part 2 example 1`] = `
Array [
  ".#",
  "#.",
]
`;
